File: apps/api/schemas/entities.py
"""
ClimaX Entity Request/Response Schemas
Defines the Pydantic DTO contracts for all 16 domain entities.
"""

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field

from schemas.ai import (
    AIExplanationSchema,
    InformationTier,
    PollutionCategory,
    VerificationStatus,
)
from schemas.base import CoordinatesDTO


class UserRole(str, Enum):
    CITIZEN = "CITIZEN"
    AUTHORITY = "AUTHORITY"
    RESEARCHER = "RESEARCHER"
    ADMIN = "ADMIN"


class IncidentStatus(str, Enum):
    OPEN = "OPEN"
    INVESTIGATING = "INVESTIGATING"
    DISPATCHED = "DISPATCHED"
    MITIGATED = "MITIGATED"
    RESOLVED = "RESOLVED"
    DISMISSED = "DISMISSED"


class IncidentSeverity(str, Enum):
    LOW = "LOW"
    MODERATE = "MODERATE"
    HIGH = "HIGH"
    VERY_HIGH = "VERY_HIGH"
    CRITICAL = "CRITICAL"


class AlertChannel(str, Enum):
    IN_APP = "IN_APP"
    SMS = "SMS"
    PUSH = "PUSH"
    EMAIL = "EMAIL"
    MUNICIPAL_BROADCAST = "MUNICIPAL_BROADCAST"


class SensorType(str, Enum):
    GOVERNMENT_STATION = "GOVERNMENT_STATION"
    IOT_LOW_COST = "IOT_LOW_COST"
    COMMUNITY = "COMMUNITY"
    SATELLITE_PIXEL = "SATELLITE_PIXEL"


# 1. User DTOs
class UserBase(BaseModel):
    email: str
    full_name: str
    role: UserRole = UserRole.CITIZEN
    organization_id: str | None = None
    preferred_language: str = "en"


class UserCreate(UserBase):
    password: str = Field(..., min_length=12, max_length=72)


class UserUpdate(BaseModel):
    full_name: str | None = Field(default=None, min_length=1, max_length=255)
    role: UserRole | None = None
    organization_id: str | None = None
    preferred_language: str | None = Field(default=None, min_length=2, max_length=10)
    is_active: bool | None = None


class UserRead(UserBase):
    id: str
    is_active: bool
    created_at: datetime


# 2. Organization DTOs
class OrganizationRead(BaseModel):
    id: str
    name: str
    jurisdiction_code: str
    department: str
    contact_email: str
    created_at: datetime


# 3. DataSource DTOs
class DataSourceRead(BaseModel):
    id: str
    name: str
    source_type: str
    provider: str
    refresh_interval_seconds: int
    is_active: bool
    created_at: datetime


# 4. Sensor DTOs
class SensorRead(BaseModel):
    id: str
    data_source_id: str
    external_sensor_id: str
    sensor_type: SensorType
    model_name: str
    location: CoordinatesDTO
    is_calibrated: bool
    is_active: bool
    last_ping_at: datetime | None = None


# 5. EnvironmentalObservation DTOs
class EnvironmentalObservationCreate(BaseModel):
    sensor_id: str
    timestamp: datetime
    pm25: float | None = None
    pm10: float | None = None
    no2: float | None = None
    so2: float | None = None
    co: float | None = None
    o3: float | None = None
    aqi: float | None = None
    temperature_c: float | None = None
    humidity_percent: float | None = None
    wind_speed_kmh: float | None = None
    wind_direction_deg: float | None = None


class EnvironmentalObservationRead(EnvironmentalObservationCreate):
    id: str
    tier: InformationTier = InformationTier.OBSERVED
    created_at: datetime


# 6. CitizenReport DTOs
class CitizenReportCreate(BaseModel):
    location: CoordinatesDTO
    address_text: str | None = None
    category: PollutionCategory
    description: str
    media_urls: list[str] = Field(default_factory=list)


class CitizenReportRead(BaseModel):
    id: str
    user_id: str | None = None
    tier: InformationTier = InformationTier.OBSERVED
    location: CoordinatesDTO
    address_text: str | None = None
    category: PollutionCategory
    description: str
    media_urls: list[str]
    status: str
    incident_id: str | None = None
    created_at: datetime


# 7. AIAnalysis DTOs
class AIAnalysisRead(BaseModel):
    id: str
    report_id: str | None = None
    observation_id: str | None = None
    tier: InformationTier = InformationTier.INFERRED
    classification: PollutionCategory
    explanation: AIExplanationSchema
    suggested_severity: IncidentSeverity
    created_at: datetime


# 8. PollutionEvent DTOs
class PollutionEventRead(BaseModel):
    id: str
    event_type: str
    tier: InformationTier
    start_time: datetime
    end_time: datetime | None = None
    peak_pm25: float | None = None
    affected_radius_meters: float
    severity: IncidentSeverity
    created_at: datetime


# 9. Prediction DTOs
class PredictionRead(BaseModel):
    id: str
    sensor_id: str | None = None
    target_location: CoordinatesDTO
    forecast_timestamp: datetime
    horizon_hours: int
    tier: InformationTier = InformationTier.PREDICTED
    predicted_pm25: float
    predicted_aqi: float
    confidence_interval_low: float
    confidence_interval_high: float
    explanation: AIExplanationSchema
    created_at: datetime


# 10. RiskAssessment DTOs
class RiskAssessmentRead(BaseModel):
    id: str
    location: CoordinatesDTO
    risk_score: float
    severity: IncidentSeverity
    population_vulnerability_index: float
    sensitive_receptors_count: int
    dominant_pollutant: str
    calculated_at: datetime


# 11. Alert DTOs
class AlertCreate(BaseModel):
    title: str
    message: str
    severity: IncidentSeverity
    channel: AlertChannel
    affected_radius_m: float | None = None


class AlertRead(AlertCreate):
    id: str
    is_dispatched: bool
    dispatched_at: datetime | None = None
    expires_at: datetime | None = None
    created_at: datetime


# 12. Incident DTOs
class IncidentCreate(BaseModel):
    title: str
    severity: IncidentSeverity
    category: PollutionCategory
    location: CoordinatesDTO
    assigned_officer_id: str | None = None


class IncidentRead(IncidentCreate):
    id: str
    organization_id: str
    status: IncidentStatus
    risk_score: float | None = None
    ai_analysis_id: str | None = None
    created_at: datetime
    resolved_at: datetime | None = None


# 13. Verification DTOs
class VerificationCreate(BaseModel):
    incident_id: str
    status: VerificationStatus
    official_notes: str
    field_photos: list[str] = Field(default_factory=list)


class VerificationRead(VerificationCreate):
    id: str
    verified_by_user_id: str
    tier: InformationTier = InformationTier.VERIFIED
    verified_at: datetime


# 14. Intervention DTOs
class InterventionCreate(BaseModel):
    incident_id: str
    intervention_type: str
    executing_agency: str
    action_summary: str


class InterventionRead(InterventionCreate):
    id: str
    dispatched_at: datetime
    executed_at: datetime | None = None
    completed_at: datetime | None = None


# 15. InterventionMeasurement DTOs
class InterventionMeasurementRead(BaseModel):
    id: str
    intervention_id: str
    pre_intervention_pm25: float
    post_intervention_pm25: float
    delta_pm25_percent: float
    evaluation_window_hours: int
    is_statistically_significant: bool
    measured_at: datetime


# 16. AuditLog DTOs
class AuditLogRead(BaseModel):
    id: str
    user_id: str | None = None
    action: str
    entity_name: str
    entity_id: str
    ip_address: str | None = None
    changes: dict[str, Any] | None = None
    timestamp: datetime

File: packages/types/src/entities.ts
/**
 * ClimaX Core Domain Entities Contract
 * Represents the unified type definitions across Frontend, Backend, and Data layers.
 */

import { AIExplanationMetadata, InformationTier, PollutionCategory, VerificationStatus } from './ai';
import { Coordinates, GeoPoint } from './geo';

export type UserRole = 'CITIZEN' | 'AUTHORITY' | 'RESEARCHER' | 'ADMIN';

export type IncidentStatus = 'OPEN' | 'INVESTIGATING' | 'DISPATCHED' | 'MITIGATED' | 'RESOLVED' | 'DISMISSED';

export type IncidentSeverity = 'LOW' | 'MODERATE' | 'HIGH' | 'VERY_HIGH' | 'CRITICAL';

export type AlertChannel = 'IN_APP' | 'SMS' | 'PUSH' | 'EMAIL' | 'MUNICIPAL_BROADCAST';

export type SensorType = 'GOVERNMENT_STATION' | 'IOT_LOW_COST' | 'COMMUNITY' | 'SATELLITE_PIXEL';

export interface User {
  id: string;
  email: string;
  full_name: string;
  role: UserRole;
  organization_id?: string;
  phone_number?: string;
  preferred_language: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface Organization {
  id: string;
  name: string;
  jurisdiction_code: string;
  department: string;
  contact_email: string;
  boundary_geojson?: string;
  created_at: string;
}

export interface DataSource {
  id: string;
  name: string;
  source_type: 'SENSOR_NETWORK' | 'GOVERNMENT_API' | 'SATELLITE' | 'CITIZEN' | 'WEATHER';
  provider: string;
  refresh_interval_seconds: number;
  is_active: boolean;
  metadata: Record<string, unknown>;
  created_at: string;
}

export interface Sensor {
  id: string;
  data_source_id: string;
  external_sensor_id: string;
  sensor_type: SensorType;
  model_name: string;
  location: Coordinates;
  geometry: GeoPoint;
  elevation_m?: number;
  is_calibrated: boolean;
  is_active: boolean;
  last_ping_at?: string;
  created_at: string;
}

export interface EnvironmentalObservation {
  id: string;
  sensor_id: string;
  timestamp: string;
  tier: InformationTier; // 'OBSERVED'
  pm25?: number;
  pm10?: number;
  no2?: number;
  so2?: number;
  co?: number;
  o3?: number;
  aqi?: number;
  temperature_c?: number;
  humidity_percent?: number;
  wind_speed_kmh?: number;
  wind_direction_deg?: number;
  raw_payload_uri?: string;
  created_at: string;
}

export interface CitizenReport {
  id: string;
  user_id?: string; // Optional for anonymous citizen reporting
  tier: InformationTier; // 'OBSERVED'
  location: Coordinates;
  geometry: GeoPoint;
  address_text?: string;
  category: PollutionCategory;
  description: string;
  media_urls: string[];
  status: 'SUBMITTED' | 'TRIAGED' | 'VERIFIED' | 'RESOLVED';
  incident_id?: string;
  created_at: string;
  updated_at: string;
}

export interface AIAnalysis {
  id: string;
  report_id?: string;
  observation_id?: string;
  tier: InformationTier; // 'INFERRED'
  classification: PollutionCategory;
  explanation: AIExplanationMetadata;
  suggested_severity: IncidentSeverity;
  plume_bounding_box?: [number, number, number, number];
  raw_model_response?: Record<string, unknown>;
  created_at: string;
}

export interface PollutionEvent {
  id: string;
  event_type: string;
  tier: InformationTier; // 'INFERRED' or 'OBSERVED'
  start_time: string;
  end_time?: string;
  geometry: GeoPoint;
  peak_pm25?: number;
  affected_radius_meters: number;
  severity: IncidentSeverity;
  created_at: string;
}

export interface Prediction {
  id: string;
  sensor_id?: string;
  target_location: Coordinates;
  forecast_timestamp: string;
  horizon_hours: number; // 6, 24, 72
  tier: InformationTier; // 'PREDICTED'
  predicted_pm25: number;
  predicted_aqi: number;
  confidence_interval_low: number;
  confidence_interval_high: number;
  explanation: AIExplanationMetadata;
  created_at: string;
}

export interface RiskAssessment {
  id: string;
  location: Coordinates;
  risk_score: number; // 0.0 to 100.0
  severity: IncidentSeverity;
  population_vulnerability_index: number;
  sensitive_receptors_count: number; // Schools, hospitals within 1km
  dominant_pollutant: string;
  calculated_at: string;
}

export interface Alert {
  id: string;
  title: string;
  message: string;
  severity: IncidentSeverity;
  channel: AlertChannel;
  target_geometry?: GeoPoint;
  affected_radius_m?: number;
  is_dispatched: boolean;
  dispatched_at?: string;
  expires_at?: string;
  created_at: string;
}

export interface Incident {
  id: string;
  organization_id: string;
  title: string;
  status: IncidentStatus;
  severity: IncidentSeverity;
  category: PollutionCategory;
  geometry: GeoPoint;
  location: Coordinates;
  assigned_officer_id?: string;
  risk_score?: number;
  ai_analysis_id?: string;
  root_cause_summary?: string;
  created_at: string;
  resolved_at?: string;
}

export interface Verification {
  id: string;
  incident_id: string;
  verified_by_user_id: string;
  tier: InformationTier; // 'VERIFIED'
  status: VerificationStatus;
  official_notes: string;
  field_photos: string[];
  verified_at: string;
}

export interface Intervention {
  id: string;
  incident_id: string;
  intervention_type: 'SMOG_GUN' | 'ROAD_WATERING' | 'FACTORY_HALT' | 'TRAFFIC_DIVERSION' | 'CLEANUP';
  dispatched_at: string;
  executed_at?: string;
  completed_at?: string;
  executing_agency: string;
  action_summary: string;
}

export interface InterventionMeasurement {
  id: string;
  intervention_id: string;
  pre_intervention_pm25: number;
  post_intervention_pm25: number;
  delta_pm25_percent: number;
  evaluation_window_hours: number;
  is_statistically_significant: boolean;
  measured_at: string;
}

export interface AuditLog {
  id: string;
  user_id?: string;
  action: string;
  entity_name: string;
  entity_id: string;
  ip_address?: string;
  changes?: Record<string, unknown>;
  timestamp: string;
}
